src/repository/default.rs

```diff
@@ -439,6 +439,54 @@ impl Repository for $repo {
         Ok(functions)
     }
 
+
+    async fn drop_function(
+        &self, database_id:
+        DatabaseId,
+        func_desc: &[sqlparser::ast::DropFunctionDesc],
+    ) -> Result<(), Error> {
+
+        for desc in func_desc.iter() {
+            let function_name = &desc.name.to_string();
+            let args = desc.args.as_ref().map(|args|
+                args.iter().map(|arg| arg.to_string()).collect::<Vec<_>>().join(", ")
+            );
+
+            // Construct the SQL DELETE statement
+            let query = if let Some(_args) = args.clone() {
+                format!(
+                    r#"
+                    DELETE FROM "function"
+                    WHERE database_id = $1 AND name = $2 AND input_types = $3;
+                    "#,
+                )
+            } else {
+                format!(
+                    r#"
+                    DELETE FROM "function"
+                    WHERE database_id = $1 AND name = $2;
+                    "#,
+                )
+            };
+
+            // Execute the SQL DELETE query 1x per function
+            let mut query = sqlx::query(&query)
+                .bind(database_id)
+                .bind(function_name);
+
+            if let Some(_args) = args.clone() {
+                query = query.bind(args);
+            }
+
+            query
```

Comment:
In SQL, running `DROP FUNCTION` for a non-existent function, e.g. running it 2x in a row, will error:

https://github.com/splitgraph/seafowl/assets/182515/041897cd-8933-4843-97a5-73b96b5412a5

Conversely, `DELETE FROM`-ing a table twice in a row does not:

https://github.com/splitgraph/seafowl/assets/182515/ee20ec8b-5fa6-438b-995b-e4ea6fac9a40


This is normally fine, but because I implemented `DROP FUNCTION` using `DELETE FROM`, running `DROP FUNCTION` 2x in a row does _not_ currently yield some kind of error, e.g. `Error::Plan`, though it seems like it probably should.

One idea I had was when we execute DELETE FROM, tally up `rows_affected` and if > 0, return Err(). 
```rust
  let rows_affected = query
                .execute(&self.executor)
                .await
                .map_err($repo::interpret_error)?
                .rows_affected();

            if rows_affected == 0 {
                return Err($SOME_ERROR);
            }
```

However, I ran into trouble importing [FunctionDoesNotExist](https://github.com/splitgraph/seafowl/pull/457/files#diff-99aeb2649cc42795a63369ffb1ba0f206d9f24bb0779b095f01e9b687481064eR48)

```rust
if rows_affected == 0 {
    return Err(crate::catalog::Error::FunctionDoesNotExist { name: function_name.to_string() });
}
```

```
error[E0308]: mismatched types
   --> src/repository/default.rs:488:28
    |
488 |                 return Err(crate::catalog::Error::FunctionDoesNotExist { name: function_name.to_string() });
    |                        --- ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^ expected `interface::Error`, found `catalog::Error`
    |                        |
    |                        arguments to this enum variant are incorrect

```

Trying to figure out an appropriate way to handle. Thoughts?

Reply:
> running DROP FUNCTION 2x in a row does not currently yield some kind of error, e.g. Error::Plan, though it seems like it probably should.

It should, conditional on the user omitting [`IF EXISTS` clause](https://github.com/sqlparser-rs/sqlparser-rs/blob/main/src/ast/mod.rs#L1407) in the `DropFunction` statement. If that clause is used then the statement shouldn't error out in either case. Here's an analogous example: https://github.com/splitgraph/seafowl/blob/main/src/context.rs#L1353-L1359

> One idea I had was when we execute DELETE FROM, tally up rows_affected and if > 0, return Err().

You're on the right track; what we usually do is add a `RETURNING id` to the `DELETE` statement, so that the DB nicely handles that check for us: https://github.com/splitgraph/seafowl/blob/main/src/repository/default.rs#L444-L454

The reason for the error above is that you're doing that check manually inside the `Repository` trait implementation which expects an [interface error](https://github.com/splitgraph/seafowl/blob/main/src/repository/interface.rs#L80-L86), while you specify the [catalog error](https://github.com/splitgraph/seafowl/blob/main/src/catalog.rs#L38-L51). By following the above prescription you can circumvent this issue.

Reply:
Thank you Marko! If I understand you correctly `if_exists` will be true, IFF the user provided `IF EXISTS` in their query. I'm adapting the rest of the logic accordingly now, thanks.

Reply:
OK, I believe these changes are implemented now. Going to mark resolved, but if any errors/changes needed, please feel free to re-open.